# Clinical Trial Data Pipeline

## Data Cleaning

Dataset: trials_master_UNCLEANED.csv

In [33]:
import pandas as pd 

# 📥 Load Dataset

In this section, we load the raw clinical trial dataset into a Pandas DataFrame for exploration and cleaning.

In [34]:
import pandas as pd

df = pd.read_csv("../Data/trials_master_UNCLEANED.csv")


In [35]:
import os

print(os.getcwd())

d:\clinical-trials-precision-oncology-analytics\notebooks


In [36]:

df.head()

,nct_id,title,matched_condition,overall_status,why_stopped,start_date_raw,completion_date_raw,lead_sponsor,sponsor_class,collaborators_raw,conditions_raw,phase_raw,study_type,enrollment_count,eligibility_criteria,minimum_age,maximum_age,sex,locations_raw
0,NCT02113852,"Study of Blood, Tumor and Adjacent Normal Tiss...",lung cancer,COMPLETED,NaN,2013-08,2016-12,Guangdong Association of Clinical Trials,OTHER,NaN,"[""Non Small Cell Lung Cancer""]",NaN,OBSERVATIONAL,250.0,Inclusion Criteria:\n\n1. Provision of written...,18 Years,NaN,ALL,NaN
1,NCT00165438,Pulmonary Function in Patients With Hodgkin's ...,lung cancer,COMPLETED,NaN,2001-10,2010-01,Dana-Farber Cancer Institute,OTHER,"[{""name"": ""Brigham and Women's Hospital"", ""cla...","[""Hodgkin's Disease""]",NaN,OBSERVATIONAL,60.0,Inclusion Criteria:\n\n* Patients with newly d...,18 Years,NaN,ALL,"[{""facility"": ""Brigham and Women's Hospital"", ..."
2,NCT02392455,VARGADO - Vargatef in 2nd-line Therapy of Non-...,lung cancer,COMPLETED,NaN,2015-03-23,2025-07-23,Boehringer Ingelheim,INDUSTRY,NaN,"[""Carcinoma, Non-Small-Cell Lung""]",NaN,OBSERVATIONAL,817.0,Inclusion criteria:\n\n* age 18 or older\n* me...,18 Years,NaN,ALL,"[{""city"": ""Multiple Locations"", ""country"": ""Ge..."
3,NCT00985855,Feasibility of Cetuximab Associated With Conco...,lung cancer,COMPLETED,NaN,2009-09,2011-08,Institut de Cancérologie de la Loire,OTHER,NaN,"[""Non-Small-Cell Lung Carcinoma""]","[""PHASE2""]",INTERVENTIONAL,91.0,Inclusion Criteria:\n\n* Non-Small-Cell Lung c...,18 Years,70 Years,ALL,"[{""facility"": ""Institut de Cancerologie de la ..."
4,NCT06074978,Early Detection of Lung Cancer In Patients Wit...,lung cancer,UNKNOWN,NaN,2023-12-20,2024-12-20,Assiut University,OTHER,NaN,"[""Chest--Diseases""]",NaN,OBSERVATIONAL,100.0,Inclusion Criteria:\n\n1- Chronic chest diseas...,45 Years,75 Years,ALL,NaN


# 🔍 Dataset Overview

Before cleaning the data, we first understand its structure, dimensions, and data types.

In [37]:
df.shape

(42829, 19)

In [38]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 42829 entries, 0 to 42828
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   nct_id                42829 non-null  str    
 1   title                 42829 non-null  str    
 2   matched_condition     42829 non-null  str    
 3   overall_status        42829 non-null  str    
 4   why_stopped           4362 non-null   str    
 5   start_date_raw        42585 non-null  str    
 6   completion_date_raw   41369 non-null  str    
 7   lead_sponsor          42829 non-null  str    
 8   sponsor_class         42829 non-null  str    
 9   collaborators_raw     15167 non-null  str    
 10  conditions_raw        42828 non-null  str    
 11  phase_raw             33935 non-null  str    
 12  study_type            42829 non-null  str    
 13  enrollment_count      42259 non-null  float64
 14  eligibility_criteria  42825 non-null  str    
 15  minimum_age           40947 no

In [39]:
df.columns 

Index(['nct_id', 'title', 'matched_condition', 'overall_status', 'why_stopped',
       'start_date_raw', 'completion_date_raw', 'lead_sponsor',
       'sponsor_class', 'collaborators_raw', 'conditions_raw', 'phase_raw',
       'study_type', 'enrollment_count', 'eligibility_criteria', 'minimum_age',
       'maximum_age', 'sex', 'locations_raw'],
      dtype='str')

## 📌 Observations

- The dataset contains **42,829 rows** and **19 columns**.
- There are **18 object (string)** columns and **1 float** column.
- The dataset contains information related to clinical trials, including study details, sponsors, eligibility criteria, enrollment, and locations.
- The **`nct_id`** column appears to be the unique identifier for each clinical trial.
- Some columns may contain missing values, which will be analyzed in the next step.

In [40]:
df.describe(include='all')

,nct_id,title,matched_condition,overall_status,why_stopped,start_date_raw,completion_date_raw,lead_sponsor,sponsor_class,collaborators_raw,conditions_raw,phase_raw,study_type,enrollment_count,eligibility_criteria,minimum_age,maximum_age,sex,locations_raw
count,42829,42829,42829,42829,4362,42585,41369,42829,42829,15167,42828,33935,42829,4.225900e+04,42825,40947,17452,42812,39043
unique,41547,41461,3,13,3196,5809,5773,7178,8,7266,17046,8,3,NaN,41328,90,113,3,30818
top,NCT02773095,Effect of Liraglutide on Blood Glucose Control...,breast cancer,COMPLETED,Slow accrual,2014-01,2026-12-31,AstraZeneca,OTHER,"[{""name"": ""National Cancer Institute (NCI)"", ""...","[""Breast Cancer""]","[""NA""]",INTERVENTIONAL,NaN,No eligibility criteria,18 Years,75 Years,ALL,"[{""facility"": ""M D Anderson Cancer Center"", ""c..."
freq,3,5,16646,21050,72,130,442,852,28118,1994,5141,11634,33939,NaN,23,31562,4147,31490,150
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.585123e+03,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.305255e+05,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000e+00,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.100000e+01,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,8.000000e+01,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.280000e+02,NaN,NaN,NaN,NaN,NaN


# 🧹 Missing Value Analysis

Before cleaning the dataset, we identify missing values to understand data completeness and determine which columns require attention.

In [41]:
df.isnull().sum()

nct_id                      0
title                       0
matched_condition           0
overall_status              0
why_stopped             38467
start_date_raw            244
completion_date_raw      1460
lead_sponsor                0
sponsor_class               0
collaborators_raw       27662
conditions_raw              1
phase_raw                8894
study_type                  0
enrollment_count          570
eligibility_criteria        4
minimum_age              1882
maximum_age             25377
sex                        17
locations_raw            3786
dtype: int64

## 📌 Missing Value Observations

- `why_stopped` contains many missing values because most studies were not terminated.
- `collaborators_raw` has many missing values, indicating that many studies have no collaborators.
- `maximum_age` has a large number of missing values, which may represent studies with no upper age limit.
- `completion_date_raw` is missing for ongoing or incomplete studies.
- `eligibility_criteria` has only 4 missing values and will be reviewed manually.
- Overall, missing values appear to be meaningful rather than random, so no columns or rows will be removed at this stage.

In [42]:
df.duplicated().sum()

np.int64(0)

In [43]:
df["nct_id"].duplicated().sum()

np.int64(1282)

In [44]:
duplicate_nct = df[df["nct_id"].duplicated(keep=False)]
duplicate_nct.sort_values("nct_id")

,nct_id,title,matched_condition,overall_status,why_stopped,start_date_raw,completion_date_raw,lead_sponsor,sponsor_class,collaborators_raw,conditions_raw,phase_raw,study_type,enrollment_count,eligibility_criteria,minimum_age,maximum_age,sex,locations_raw
3735,NCT00001805,A Phase II Clinical Trial of Suppression of Hu...,lung cancer,COMPLETED,NaN,1999-03,2000-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Neoplasms"", ""Colonic Neoplasms"", ""Lun...","[""PHASE2""]",INTERVENTIONAL,20.0,Patients must have advanced stage solid tumor ...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
17440,NCT00001805,A Phase II Clinical Trial of Suppression of Hu...,breast cancer,COMPLETED,NaN,1999-03,2000-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Neoplasms"", ""Colonic Neoplasms"", ""Lun...","[""PHASE2""]",INTERVENTIONAL,20.0,Patients must have advanced stage solid tumor ...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
17992,NCT00001944,Vinorelbine and XR9576 to Treat Cancer,breast cancer,COMPLETED,NaN,1999-12,2001-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Cancer"", ""Cancer"", ""Lung Cancer"", ""Ov...","[""PHASE1""]",INTERVENTIONAL,30.0,Age greater than or equal to 18 years.\n\nHist...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
10323,NCT00001944,Vinorelbine and XR9576 to Treat Cancer,lung cancer,COMPLETED,NaN,1999-12,2001-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Cancer"", ""Cancer"", ""Lung Cancer"", ""Ov...","[""PHASE1""]",INTERVENTIONAL,30.0,Age greater than or equal to 18 years.\n\nHist...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
18810,NCT00002465,High-Dose Megestrol in Treating Patients With ...,breast cancer,UNKNOWN,NaN,1987-12,NaN,St. Vincent Medical Center - Los Angeles,OTHER,NaN,"[""Breast Cancer"", ""Endometrial Cancer"", ""Malig...","[""PHASE1"", ""PHASE2""]",INTERVENTIONAL,NaN,DISEASE CHARACTERISTICS: Histologically docume...,18 Years,NaN,FEMALE,"[{""facility"": ""St. Vincent Medical Center - Lo..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21345,NCT07689721,ctDNA-Guided Intraventricular Therapy for CNS ...,breast cancer,NOT_YET_RECRUITING,NaN,2027-01-01,2032-01-01,Alireza Mansouri,OTHER,NaN,"[""Leptomeningeal Disease"", ""Glioblastoma"", ""Br...","[""NA""]",INTERVENTIONAL,77.0,Inclusion Criteria:\n\n* Age ≥ 18 years.\n* Di...,18 Years,120 Years,ALL,NaN
8232,NCT07705074,FAPI PET Imaging - An Exploratory Study,lung cancer,NOT_YET_RECRUITING,NaN,2026-08,2029-09,Martin Huellner,OTHER,"[{""name"": ""University Hospital, Z\u00fcrich"", ...","[""Thyroid Pathology"", ""Head and Neck Cancer (H...","[""PHASE1"", ""PHASE2""]",INTERVENTIONAL,770.0,Inclusion Criteria:\n\n* Written informed cons...,18 Years,NaN,ALL,"[{""facility"": ""Kantonsspital Baden"", ""city"": ""..."
17262,NCT07705074,FAPI PET Imaging - An Exploratory Study,breast cancer,NOT_YET_RECRUITING,NaN,2026-08,2029-09,Martin Huellner,OTHER,"[{""name"": ""University Hospital, Z\u00fcrich"", ...","[""Thyroid Pathology"", ""Head and Neck Cancer (H...","[""PHASE1"", ""PHASE2""]",INTERVENTIONAL,770.0,Inclusion Criteria:\n\n* Written informed cons...,18 Years,NaN,ALL,"[{""facility"": ""Kantonsspital Baden"", ""city"": ""..."
11755,NCT07705334,"A Study of SHY-ONC6, a Novel Proteasome Inhibi...",lung cancer,RECRUITING,NaN,2026-06-29,2028-05-15,SHY Therapeutics,INDUSTRY,NaN,"[""Advanced or Metastatic Solid Tumors"", ""Tripl...","[""PHASE1""]",INTERVENTIONAL,30.0,Inclusion Criteria:\n\n* Male or female ≥18 ye...,18 Years,NaN,ALL,"[{""facility"": ""SCRI at HCA HealthONE"", ""status..."


In [45]:
duplicate_nct.sort_values("nct_id").head(10)

,nct_id,title,matched_condition,overall_status,why_stopped,start_date_raw,completion_date_raw,lead_sponsor,sponsor_class,collaborators_raw,conditions_raw,phase_raw,study_type,enrollment_count,eligibility_criteria,minimum_age,maximum_age,sex,locations_raw
3735,NCT00001805,A Phase II Clinical Trial of Suppression of Hu...,lung cancer,COMPLETED,NaN,1999-03,2000-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Neoplasms"", ""Colonic Neoplasms"", ""Lun...","[""PHASE2""]",INTERVENTIONAL,20.0,Patients must have advanced stage solid tumor ...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
17440,NCT00001805,A Phase II Clinical Trial of Suppression of Hu...,breast cancer,COMPLETED,NaN,1999-03,2000-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Neoplasms"", ""Colonic Neoplasms"", ""Lun...","[""PHASE2""]",INTERVENTIONAL,20.0,Patients must have advanced stage solid tumor ...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
17992,NCT00001944,Vinorelbine and XR9576 to Treat Cancer,breast cancer,COMPLETED,NaN,1999-12,2001-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Cancer"", ""Cancer"", ""Lung Cancer"", ""Ov...","[""PHASE1""]",INTERVENTIONAL,30.0,Age greater than or equal to 18 years.\n\nHist...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
10323,NCT00001944,Vinorelbine and XR9576 to Treat Cancer,lung cancer,COMPLETED,NaN,1999-12,2001-06,National Cancer Institute (NCI),NIH,NaN,"[""Breast Cancer"", ""Cancer"", ""Lung Cancer"", ""Ov...","[""PHASE1""]",INTERVENTIONAL,30.0,Age greater than or equal to 18 years.\n\nHist...,NaN,NaN,ALL,"[{""facility"": ""National Cancer Institute (NCI)..."
18810,NCT00002465,High-Dose Megestrol in Treating Patients With ...,breast cancer,UNKNOWN,NaN,1987-12,NaN,St. Vincent Medical Center - Los Angeles,OTHER,NaN,"[""Breast Cancer"", ""Endometrial Cancer"", ""Malig...","[""PHASE1"", ""PHASE2""]",INTERVENTIONAL,NaN,DISEASE CHARACTERISTICS: Histologically docume...,18 Years,NaN,FEMALE,"[{""facility"": ""St. Vincent Medical Center - Lo..."
13710,NCT00002465,High-Dose Megestrol in Treating Patients With ...,lung cancer,UNKNOWN,NaN,1987-12,NaN,St. Vincent Medical Center - Los Angeles,OTHER,NaN,"[""Breast Cancer"", ""Endometrial Cancer"", ""Malig...","[""PHASE1"", ""PHASE2""]",INTERVENTIONAL,NaN,DISEASE CHARACTERISTICS: Histologically docume...,18 Years,NaN,FEMALE,"[{""facility"": ""St. Vincent Medical Center - Lo..."
10291,NCT00002475,Cyclophosphamide Plus Vaccine Therapy in Treat...,lung cancer,COMPLETED,NaN,1991-04,2009-06,St. Vincent Medical Center - Los Angeles,OTHER,NaN,"[""Breast Cancer"", ""Colorectal Cancer"", ""Kidney...","[""PHASE2""]",INTERVENTIONAL,40.0,DISEASE CHARACTERISTICS:\n\n* Histologically c...,18 Years,NaN,ALL,"[{""facility"": ""St. Vincent Medical Center - Lo..."
28966,NCT00002475,Cyclophosphamide Plus Vaccine Therapy in Treat...,breast cancer,COMPLETED,NaN,1991-04,2009-06,St. Vincent Medical Center - Los Angeles,OTHER,NaN,"[""Breast Cancer"", ""Colorectal Cancer"", ""Kidney...","[""PHASE2""]",INTERVENTIONAL,40.0,DISEASE CHARACTERISTICS:\n\n* Histologically c...,18 Years,NaN,ALL,"[{""facility"": ""St. Vincent Medical Center - Lo..."
24429,NCT00002520,Physician-Initiated Stop-Smoking Program for P...,breast cancer,COMPLETED,NaN,1990-12-07,NaN,Eastern Cooperative Oncology Group,NETWORK,"[{""name"": ""National Cancer Institute (NCI)"", ""...","[""Bladder Cancer"", ""Breast Cancer"", ""Colorecta...","[""NA""]",INTERVENTIONAL,434.0,DISEASE CHARACTERISTICS:\n\n* One of the follo...,19 Years,120 Years,ALL,"[{""facility"": ""Fox Chase Cancer Center"", ""city..."
1387,NCT00002520,Physician-Initiated Stop-Smoking Program for P...,lung cancer,COMPLETED,NaN,1990-12-07,NaN,Eastern Cooperative Oncology Group,NETWORK,"[{""name"": ""National Cancer Institute (NCI)"", ""...","[""Bladder Cancer"", ""Breast Cancer"", ""Colorecta...","[""NA""]",INTERVENTIONAL,434.0,DISEASE CHARACTERISTICS:\n\n* One of the follo...,19 Years,120 Years,ALL,"[{"

In [46]:
df[df["nct_id"] == "NCT00001805"].T

,3735,17440
nct_id,NCT00001805,NCT00001805
title,A Phase II Clinical Trial of Suppression of Hu...,A Phase II Clinical Trial of Suppression of Hu...
matched_condition,lung cancer,breast cancer
overall_status,COMPLETED,COMPLETED
why_stopped,NaN,NaN
start_date_raw,1999-03,1999-03
completion_date_raw,2000-06,2000-06
lead_sponsor,National Cancer Institute (NCI),National Cancer Institute (NCI)
sponsor_class,NIH,NIH
collaborators_raw,NaN,NaN


In [47]:
df[df["nct_id"]=="NCT00001805"]["matched_condition"]

3735       lung cancer
17440    breast cancer
Name: matched_condition, dtype: str

## 📌 Duplicate NCT ID Investigation

Although 1,282 duplicate `nct_id` values were detected, manual inspection showed that these records were **not accidental duplicates**.

The repeated records represent the same clinical trial associated with different `matched_condition` values (for example, lung cancer and breast cancer).

Since this preserves disease-specific mappings, these records were retained and no duplicate `nct_id` rows were removed.

In [48]:
df.dtypes

nct_id                      str
title                       str
matched_condition           str
overall_status              str
why_stopped                 str
start_date_raw              str
completion_date_raw         str
lead_sponsor                str
sponsor_class               str
collaborators_raw           str
conditions_raw              str
phase_raw                   str
study_type                  str
enrollment_count        float64
eligibility_criteria        str
minimum_age                 str
maximum_age                 str
sex                         str
locations_raw               str
dtype: object

# 🔧 Data Type Inspection

Before performing transformations, the data types of all columns are inspected to identify columns that require conversion into appropriate formats.

In [49]:
df["enrollment_count"].describe()

count    4.225900e+04
mean     3.585123e+03
std      1.305255e+05
min      0.000000e+00
25%      3.100000e+01
50%      8.000000e+01
75%      2.280000e+02
max      1.500000e+07
Name: enrollment_count, dtype: float64

In [50]:
df["start_date"] = pd.to_datetime(df["start_date_raw"], errors="coerce")

df["completion_date"] = pd.to_datetime(df["completion_date_raw"], errors="coerce")

In [51]:
df[["start_date_raw", "start_date",
    "completion_date_raw", "completion_date"]].head()

,start_date_raw,start_date,completion_date_raw,completion_date
0,2013-08,2013-08-01,2016-12,2016-12-01
1,2001-10,2001-10-01,2010-01,2010-01-01
2,2015-03-23,NaT,2025-07-23,NaT
3,2009-09,2009-09-01,2011-08,2011-08-01
4,2023-12-20,NaT,2024-12-20,NaT


## 📌 Data Type Observations

- Most columns were imported as text (`object`) from the CSV file.
- The `enrollment_count` column was correctly recognized as a numeric (`float64`) column.
- The date columns contained mixed formats (`YYYY-MM` and `YYYY-MM-DD`), so they were converted using `pd.to_datetime()` with `format="mixed"` to correctly parse both formats.

In [52]:
df["start_date"] = pd.to_datetime(
    df["start_date_raw"],
    format="mixed",
    errors="coerce"
)

df["completion_date"] = pd.to_datetime(
    df["completion_date_raw"],
    format="mixed",
    errors="coerce"
)

In [53]:
df[[
    "start_date_raw",
    "start_date",
    "completion_date_raw",
    "completion_date"
]].head()

,start_date_raw,start_date,completion_date_raw,completion_date
0,2013-08,2013-08-01,2016-12,2016-12-01
1,2001-10,2001-10-01,2010-01,2010-01-01
2,2015-03-23,2015-03-23,2025-07-23,2025-07-23
3,2009-09,2009-09-01,2011-08,2011-08-01
4,2023-12-20,2023-12-20,2024-12-20,2024-12-20


# 🧹 Text Standardization

This step removes unnecessary whitespace from text columns to improve consistency and prevent duplicate values caused by formatting differences.

In [54]:
df.select_dtypes(include = "object").columns

C:\Users\HP\AppData\Local\Temp\ipykernel_69832\3078658240.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.select_dtypes(include = "object").columns


Index(['nct_id', 'title', 'matched_condition', 'overall_status', 'why_stopped',
       'start_date_raw', 'completion_date_raw', 'lead_sponsor',
       'sponsor_class', 'collaborators_raw', 'conditions_raw', 'phase_raw',
       'study_type', 'eligibility_criteria', 'minimum_age', 'maximum_age',
       'sex', 'locations_raw'],
      dtype='str')

In [55]:
text_columns = df.select_dtypes(include = "object").columns

C:\Users\HP\AppData\Local\Temp\ipykernel_69832\2340328551.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_columns = df.select_dtypes(include = "object").columns


In [56]:
df[text_columns] = df[text_columns].apply(
    lambda x: x.str.strip() if x.dtype == "object" else x
)

In [57]:
df.head()

,nct_id,title,matched_condition,overall_status,why_stopped,start_date_raw,completion_date_raw,lead_sponsor,sponsor_class,collaborators_raw,...,phase_raw,study_type,enrollment_count,eligibility_criteria,minimum_age,maximum_age,sex,locations_raw,start_date,completion_date
0,NCT02113852,"Study of Blood, Tumor and Adjacent Normal Tiss...",lung cancer,COMPLETED,NaN,2013-08,2016-12,Guangdong Association of Clinical Trials,OTHER,NaN,...,NaN,OBSERVATIONAL,250.0,Inclusion Criteria:\n\n1. Provision of written...,18 Years,NaN,ALL,NaN,2013-08-01,2016-12-01
1,NCT00165438,Pulmonary Function in Patients With Hodgkin's ...,lung cancer,COMPLETED,NaN,2001-10,2010-01,Dana-Farber Cancer Institute,OTHER,"[{""name"": ""Brigham and Women's Hospital"", ""cla...",...,NaN,OBSERVATIONAL,60.0,Inclusion Criteria:\n\n* Patients with newly d...,18 Years,NaN,ALL,"[{""facility"": ""Brigham and Women's Hospital"", ...",2001-10-01,2010-01-01
2,NCT02392455,VARGADO - Vargatef in 2nd-line Therapy of Non-...,lung cancer,COMPLETED,NaN,2015-03-23,2025-07-23,Boehringer Ingelheim,INDUSTRY,NaN,...,NaN,OBSERVATIONAL,817.0,Inclusion criteria:\n\n* age 18 or older\n* me...,18 Years,NaN,ALL,"[{""city"": ""Multiple Locations"", ""country"": ""Ge...",2015-03-23,2025-07-23
3,NCT00985855,Feasibility of Cetuximab Associated With Conco...,lung cancer,COMPLETED,NaN,2009-09,2011-08,Institut de Cancérologie de la Loire,OTHER,NaN,...,"[""PHASE2""]",INTERVENTIONAL,91.0,Inclusion Criteria:\n\n* Non-Small-Cell Lung c...,18 Years,70 Years,ALL,"[{""facility"": ""Institut de Cancerologie de la ...",2009-09-01,2011-08-01
4,NCT06074978,Early Detection of Lung Cancer In Patients Wit...,lung cancer,UNKNOWN,NaN,2023-12-20,2024-12-20,Assiut University,OTHER,NaN,...,NaN,OBSERVATIONAL,100.0,Inclusion Criteria:\n\n1- Chronic chest diseas...,45 Years,75 Years,ALL,NaN,2023-12-20,2024-12-20


## 📌 Observation

Leading and trailing whitespaces were removed from all text columns to ensure consistent formatting and avoid mismatched values during analysis.

# 🔍 Missing Value Investigation

Missing values were analyzed column by column to understand whether they represent missing information, non-applicable values, or expected characteristics of clinical trial data.


In [58]:
df["overall_status"].value_counts() 

overall_status
COMPLETED                    21050
UNKNOWN                       6336
RECRUITING                    5627
TERMINATED                    3529
ACTIVE_NOT_RECRUITING         2544
NOT_YET_RECRUITING            2011
WITHDRAWN                     1226
ENROLLING_BY_INVITATION        266
SUSPENDED                      154
NO_LONGER_AVAILABLE             36
APPROVED_FOR_MARKETING          26
AVAILABLE                       21
TEMPORARILY_NOT_AVAILABLE        3
Name: count, dtype: int64

## 📌 Observation

The `overall_status` column contains multiple trial states such as **Completed**, **Recruiting**, **Terminated**, and **Suspended**.

The high number of missing values in the `why_stopped` column is expected because trials that are completed or still ongoing do not require a stopping reason. Therefore, missing values in `why_stopped` will be retained rather than treated as data quality issues.

In [59]:
df["phase_raw"].value_counts(dropna=False)

phase_raw
["NA"]                  11634
NaN                      8894
["PHASE2"]               8466
["PHASE1"]               4665
["PHASE3"]               4087
["PHASE1", "PHASE2"]     2196
["PHASE4"]               1977
["PHASE2", "PHASE3"]      460
["EARLY_PHASE1"]          450
Name: count, dtype: int64

In [60]:
df[df["phase_raw"].isna()]["study_type"].value_counts()

study_type
OBSERVATIONAL      8804
EXPANDED_ACCESS      86
INTERVENTIONAL        4
Name: count, dtype: int64

In [61]:
df[df["phase_raw"] == '["NA"]']["study_type"].value_counts()

study_type
INTERVENTIONAL    11634
Name: count, dtype: int64

## 📌 Observation

Investigation of the `phase_raw` column showed two different representations of unavailable phase information:

- `NaN` values were primarily associated with **Observational** studies, where clinical phases are generally not applicable.
- The string value `["NA"]` was found in **Interventional** studies and was retained because it represents an explicit value present in the source dataset rather than a missing entry.

Therefore, no modifications were made to the `phase_raw` column during data cleaning.

In [62]:
df["sex"].value_counts(dropna=False)

sex
ALL       31490
FEMALE    10795
MALE        527
NaN          17
Name: count, dtype: int64

## 📌 Observation

The `sex` column contains only 17 missing values out of 42,829 records (approximately 0.04% of the dataset).

Since the correct participant eligibility cannot be inferred, these missing values were retained to preserve data integrity.

In [63]:
df["minimum_age"].value_counts(dropna=False)

minimum_age
18 Years     31562
NaN           1882
20 Years      1612
40 Years      1105
30 Years      1060
             ...  
30 Days          1
54 Years         1
38 Months        1
3 Days           1
40 Weeks         1
Name: count, Length: 91, dtype: int64

In [64]:
df["maximum_age"].value_counts(dropna=False)

maximum_age
NaN          25377
75 Years      4147
70 Years      2782
80 Years      2288
65 Years      1905
             ...  
8 Days           1
26 Years         1
97 Years         1
115 Years        1
8 Years          1
Name: count, Length: 114, dtype: int64

## 📌 Observation

The `minimum_age` column contains a small number of missing values, while the `maximum_age` column has a large proportion of missing entries.

These missing values were retained because they likely represent studies where age limits were not specified rather than errors in the dataset. No imputation was performed to avoid introducing incorrect information.

# ✅ Data Validation

After cleaning the dataset, validation checks were performed to identify impossible or inconsistent values that could affect the accuracy of the analysis.

In [65]:
(df["enrollment_count"]<0).sum()

np.int64(0)

## 📌 Observation

The `enrollment_count` column was checked for negative values.

No negative enrollment counts were found, indicating that the participant counts are logically valid.

In [66]:
(df['completion_date'] < df["start_date"]).sum() 

np.int64(0)

## 📌 Observation

The start and completion dates were validated to ensure chronological consistency.

No records were found where the completion date occurred before the start date, indicating that the dataset does not contain logically inconsistent study timelines.

In [69]:
df['minimum_age'].head(10)

0    18 Years
1    18 Years
2    18 Years
3    18 Years
4    45 Years
5    15 Years
6    30 Years
7    55 Years
8    18 Years
9    18 Years
Name: minimum_age, dtype: str

In [70]:
df['maximum_age'].head(10)

0         NaN
1         NaN
2         NaN
3    70 Years
4    75 Years
5         NaN
6    70 Years
7    77 Years
8    75 Years
9         NaN
Name: maximum_age, dtype: str

## 📌 Observation

The `minimum_age` and `maximum_age` columns are stored as text values (e.g., "18 Years") rather than numeric values.

To preserve the original dataset, these columns were retained in their original format. Numeric age columns can be created later if age-based statistical analysis is required.

In [71]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 42829 entries, 0 to 42828
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   nct_id                42829 non-null  str           
 1   title                 42829 non-null  str           
 2   matched_condition     42829 non-null  str           
 3   overall_status        42829 non-null  str           
 4   why_stopped           4362 non-null   str           
 5   start_date_raw        42585 non-null  str           
 6   completion_date_raw   41369 non-null  str           
 7   lead_sponsor          42829 non-null  str           
 8   sponsor_class         42829 non-null  str           
 9   collaborators_raw     15167 non-null  str           
 10  conditions_raw        42828 non-null  str           
 11  phase_raw             33935 non-null  str           
 12  study_type            42829 non-null  str           
 13  enrollment_count      42259

In [72]:
df.duplicated().sum()

np.int64(0)

In [73]:
df.to_csv("trials_master_CLEANED.csv", index=False)

In [74]:
import os

os.path.exists("trials_master_CLEANED.csv")

True

In [76]:
import os

os.getcwd()

'd:\\clinical-trials-precision-oncology-analytics\\notebooks'

In [77]:
sample_df.to_csv("../data/sample_data.csv", index=False)

In [78]:
import os

os.path.exists("../data/sample_data.csv")

True